File: src-tauri/src/parser/discovery.rs
use std::path::PathBuf;
use std::fs;
use walkdir::WalkDir;

#[derive(Debug, Clone)]
pub struct SessionFile {
    pub path: PathBuf,
    pub dir_name: String,
    pub file_name: String,
    pub file_size: u64,
}

pub fn discover_session_files() -> Result<Vec<SessionFile>, String> {
    let home_dir = dirs::home_dir()
        .ok_or_else(|| "Could not determine home directory".to_string())?;
    
    let sessions_dir = home_dir.join(".pi").join("agent").join("sessions");
    
    if !sessions_dir.exists() {
        return Ok(Vec::new());
    }

    let mut session_files = Vec::new();

    // Walk each subdirectory in the sessions directory
    for entry in fs::read_dir(&sessions_dir)
        .map_err(|e| format!("Failed to read sessions directory: {}", e))? 
    {
        let entry = entry.map_err(|e| format!("Failed to read directory entry: {}", e))?;
        let path = entry.path();
        
        if !path.is_dir() {
            continue;
        }

        let dir_name = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_string();

        // Walk all .jsonl files in this project directory
        for file_entry in WalkDir::new(&path)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter(|e| e.file_type().is_file())
        {
            let file_path = file_entry.path();
            
            if let Some(extension) = file_path.extension() {
                if extension == "jsonl" {
                    let file_name = file_path
                        .file_name()
                        .and_then(|n| n.to_str())
                        .unwrap_or("")
                        .to_string();

                    let file_size = file_entry
                        .metadata()
                        .map(|m| m.len())
                        .unwrap_or(0);

                    session_files.push(SessionFile {
                        path: file_path.to_path_buf(),
                        dir_name: dir_name.clone(),
                        file_name,
                        file_size,
                    });
                }
            }
        }
    }

    // Sort by file name for consistent ordering
    session_files.sort_by(|a, b| a.file_name.cmp(&b.file_name));

    Ok(session_files)
}

File: backend/analytics/discovery.ts
/**
 * Session file discovery — faithful port of
 * the session-file discovery logic.
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

import type { SessionFile } from "./session-types";

/**
 * Walk a directory recursively, collecting all file paths.
 */
function walk_dir_recursive(dir: string): string[] {
	const results: string[] = [];
	const entries = fs.readdirSync(dir, { withFileTypes: true });

	for (const entry of entries) {
		const full_path = path.join(dir, entry.name);
		if (entry.isDirectory()) {
			results.push(...walk_dir_recursive(full_path));
		} else if (entry.isFile()) {
			results.push(full_path);
		}
	}

	return results;
}

/**
 * Discover all .jsonl session files under the sessions root directory.
 *
 * Root is resolved from `ARIADNE_PI_SESSIONS_ROOT` env var, falling back
 * to `~/.pi/agent/sessions`.
 *
 * Returns files sorted alphabetically by file_name.
 */
export function discover_session_files(): SessionFile[] {
	const sessions_dir =
		process.env.ARIADNE_PI_SESSIONS_ROOT ??
		path.join(os.homedir(), ".pi", "agent", "sessions");

	if (!fs.existsSync(sessions_dir)) {
		return [];
	}

	const session_files: SessionFile[] = [];

	// Read first-level subdirectories (URL-encoded project paths)
	let top_entries: fs.Dirent[];
	try {
		top_entries = fs.readdirSync(sessions_dir, { withFileTypes: true });
	} catch {
		return [];
	}

	for (const top_entry of top_entries) {
		const top_path = path.join(sessions_dir, top_entry.name);

		if (!top_entry.isDirectory()) {
			continue;
		}

		const dir_name = top_entry.name;

		// Walk all files recursively within this project directory
		let all_files: string[];
		try {
			all_files = walk_dir_recursive(top_path);
		} catch {
			continue;
		}

		for (const file_path of all_files) {
			if (path.extname(file_path) !== ".jsonl") {
				continue;
			}

			const file_name = path.basename(file_path);

			let file_size = 0;
			try {
				const stat = fs.statSync(file_path);
				file_size = stat.size;
			} catch {
				// default to 0 on error, matching Rust unwrap_or(0)
			}

			session_files.push({
				path: file_path,
				dir_name,
				file_name,
				file_size,
			});
		}
	}

	// Sort by file_name alphabetically
	session_files.sort((a, b) => a.file_name.localeCompare(b.file_name));

	return session_files;
}
